src/data-fetchers/vcf/vcf-worker.ts

```diff
@@ -21,25 +21,98 @@ type VcfFileOptions = {
+            this.#parser = new VCF({ header });
+        }
+        return this.#parser;
+    }
+    /**
+     * Retrieves data within a certain coordinate range
+     * @param minX A number which is the minimum X boundary of the tile
+     * @param maxX A number which is the maximum X boundary of the tile
+     * @param callback A callback function which receives the VCF tile
+     * @returns A promise of array of promises which resolve when the data has been successfully retrieved
+     */
+    async getTileData(minX: number, maxX: number, callback: (tileRecord: VcfTile) => unknown) {
+        const source = dataSources.get(this.#uid)!;
+        const parser = await this.getParser();
+
+        let curMinX = minX;
+        const { chromLengths, cumPositions } = source.chromInfo;
+        const recordPromises: Promise<void>[] = [];
+
+        cumPositions.forEach(cumPos => {
+            const chromName = cumPos.chr;
+            const chromStart = cumPos.pos;
+            const chromEnd = cumPos.pos + chromLengths[chromName];
+            let startPos, endPos;
+            if (chromStart <= curMinX && curMinX < chromEnd) {
+                // start of the visible region is within this chromosome
+                let prevPOS: number | undefined;
+                if (maxX > chromEnd) {
+                    // the visible region extends beyond the end of this chromosome
+                    // fetch from the start until the end of the chromosome
+
+                    startPos = curMinX - chromStart;
+                    endPos = chromEnd - chromStart;
+                    recordPromises.push(
+                        source.file.tbi
+                            .getLines(chromName, startPos, endPos, line => {
+                                const vcfRecord = parser.parseLine(line);
+                                const vcfTile = recordToTile(vcfRecord, chromStart, prevPOS);
+                                prevPOS = vcfTile.POS;
+                                callback(vcfTile);
+                            })
+                            .then(() => {})
+                    );
+                } else {
+                    startPos = Math.floor(curMinX - chromStart);
+                    endPos = Math.ceil(maxX - chromStart);
+                    recordPromises.push(
+                        source.file.tbi
+                            .getLines(chromName, startPos, endPos, line => {
+                                const vcfRecord = parser.parseLine(line);
+                                const vcfTile = recordToTile(vcfRecord, chromStart, prevPOS);
+                                prevPOS = vcfTile.POS;
+                                callback(vcfTile);
+                            })
+                            .then(() => {})
+                    );
+                    return;
+                }
+
+                curMinX = chromEnd;
+            }
+        });
+        return recordPromises;
```

Comment:
This is great! Although, if the final goal is to promisify the get lines callback, then I wonder if we can have `getTileData(minX, maxX): Promise<VcfTile>`? Untested code (with this idea):

```typescript
// ...
    async getTileData(minX: number, maxX: number) {
        const source = dataSources.get(this.#uid)!;
        const parser = await this.getParser();

        let curMinX = minX;
        const { chromLengths, cumPositions } = source.chromInfo;
        const recordPromises: Promise<VcfTile[]>[] = [];

        for (const cumPos of cumPositions) {
            const chromName = cumPos.chr;
            const chromStart = cumPos.pos;
            const chromEnd = cumPos.pos + chromLengths[chromName];
            let startPos, endPos;

   		    // Early break, rather than creating an nested if
            if (chromStart > curMinX || curMinX >= chromEnd) {
                continue;
            }

            // start of the visible region is within this chromosome
            let prevPOS: number | undefined;
            const tilesPromise = new Promise<VcfTile[]>(resolve => {
                const tiles: VcfTile[] = [];
                const lineCallback = (line: string) => {
                    const vcfRecord = parser.parseLine(line);
                    const vcfTile = recordToTile(vcfRecord, chromStart, prevPOS);
                    prevPOS = vcfTile.POS;
                    tiles.push(vcfTile);
                };

                if (maxX > chromEnd) {
                    startPos = curMinX - chromStart;
                    endPos = chromEnd - chromStart;
                } else {
                    startPos = Math.floor(curMinX - chromStart);
                    endPos = Math.ceil(maxX - chromStart);
                }

                source.file.tbi.getLines(chromName, startPos, endPos, lineCallback).then(() => {
                    resolve(tiles);
                });
            });

            recordPromises.push(tilesPromise);

            if (maxX <= chromEnd) {
                continue;
            }

            curMinX = chromEnd;
        }

        const tileArrays = await Promise.all(recordPromises);
        return tileArrays.flat();
    }

```

Reply:
Then, in the caller we just have:

```typescript
    tileValues[CACHE_KEY] = await source.file.getTileData(minX, maxX);
    return tileValues[CACHE_KEY];
```

Reply:
This is SO much cleaner! I forgot that `getLines` returned a promise---resolving the tiles after that makes a lot of sense. And definitely my bad for mixing callbacks and promises, will try to avoid. 

I'll refactor the bed fetcher to resemble this one.